Initial example of longevity test

vinayakshnd opened the pull request:
-  This framework allows user to trigger various testTriggers at a frequency controlled by chaos level for that particular trigger. 

-  Create a kubernetes configMap named `longevity-triggers` in `default` namespace. 
This configMap contains: 
     1. Chaos level for each test trigger. Keys in configMap can be referred from [here](https://github.com/portworx/torpedo/pull/505/files#diff-6a7206932fa1a52521aaeee0bac5d90eR47) and chaos levels from [here](https://github.com/portworx/torpedo/pull/505/files#diff-16f71344b6650f3694eff00e811382c6R220). Chaos level of 0 means the test trigger is disabled.
     2. A comma-separated list of email IDs to which send email notification at configurable interval

 For e.g 

```
Name:         longevity-triggers
Namespace:    default
Labels:       <none>
Annotations:  <none>

Data
====
emailRecipients:
----
test@portworx.com,eng@portworx.com
emailReporter:
----
9
haUpdate:
----
10
rebootNode:
----
0
restartVolDriver:
----
0
crashVolDriver:
----
0
deleteApp:
----
0
Events:  <none>
```





- Private Jenkins Job:
https://jenkins.portworx.co/job/Torpedo/job/tp-vinayak-longevity/

harsh-px commented:
@Rohit-PX can you do one pass at this and check how it compares to the longevity in px-test?

mborodin-px commented on tests/longevity/longevity_test.go:
```diff
@@ -0,0 +1,342 @@
+	configMapNS           = "default"
+)
+
+type TestTrigger string
```
unused type 'TestTrigger'

mborodin-px commented on tests/longevity/longevity_test.go:
```diff
@@ -0,0 +1,342 @@
+		logrus.Warnf("Chaos level for trigger [%s] not found in chaos map. Using global chaos level [%d]",
+			triggerType, Inst().ChaosLevel)
+	}
+	return (triggerInterval[triggerType][chaosLevel]), nil
```
remove brackets

thiguetta commented:
@vinayakshnd before merging this please run a tp-dev-test on a regular path without longevity just to see if it's not affecting the other workflows

vinayakshnd commented:
> @vinayakshnd before merging this please run a tp-dev-test on a regular path without longevity just to see if it's not affecting the other workflows

tp-dev-test job passed with this PR:
https://jenkins.portworx.co/job/Torpedo/job/tp-dev-test/742/

vinayakshnd commented:
Latest tp-dev-test run:
https://jenkins.portworx.dev/job/Torpedo/job/tp-dev-test/765/console

Latest longevity run:
https://jenkins.portworx.dev/job/Torpedo/job/tp-vinayak-longevity/73/console

Mixed triggers are getting invoked now:

**appTaskDown**: every 3 mins
**haUpdate**: every 10 mins
**Email notification**: every 20 mins
(All configurable)

<img width="1003" alt="Screenshot 2020-09-02 at 5 38 49 PM" src="https://user-images.githubusercontent.com/4953411/91979467-76cd9e00-ed43-11ea-9af9-a86e24bee0bb.png">

harsh-px commented on tests/longevity/longevity_test.go:
```diff
@@ -0,0 +1,340 @@
+				triggerFunc(contexts, triggerEventsChan)
+
+				if isDisruptiveTrigger(triggerType) {
+					triggerLoc.Unlock()
```
make this `defer triggerLoc.Unlock()`

harsh-px commented on tests/longevity/longevity_test.go:
```diff
@@ -0,0 +1,340 @@
+				lastInvocationTime = time.Now().Local()
+			}
+		}
+		time.Sleep(time.Second * 15)
```
use const for time

harsh-px commented on tests/longevity/longevity_test.go:
```diff
@@ -0,0 +1,340 @@
+	} else {
+		// If trigger is non-disruptive then just check if no other disruptive trigger is running or not
+		// and release the lock immidiately so that other non-disruptive triggers can happen.
+		triggerLoc.Lock()
+		triggerLoc.Unlock()
```
remove

vinayakshnd replied:
@harsh-px this is required, for non-disruptive triggers, they need to check if any other disruptive triggers are happening or not. If it is able to acquire the lock here that means there no other disruptive event happening at this time. Then it can release the lock and continue ahead.

harsh-px commented on tests/longevity/longevity_test.go:
```diff
@@ -0,0 +1,342 @@
+	// Stores mapping between chaos level and its freq. Values are hardcoded
+	triggerInterval map[string]map[int]time.Duration
+	// Stores which are disruptive triggers. When disruptive triggers are happening in test,
+	// other triggers are allowed to happen only after existing triggers are completed.
```
nit: `after existing disruptive triggers are complete`

harsh-px commented on tests/longevity/longevity_test.go:
```diff
@@ -0,0 +1,342 @@
+		AppTaskDown:      TriggerAppTaskDown,
+	}
+	It("has to schedule app and introduce test triggers", func() {
+		Step(fmt.Sprintf("Start watch on K8S configMap [%s] in namespace [%s]",
```
nit: change to `... K8S configMap [%s/%s]", configMapNS , testTriggersConfigMap `

harsh-px commented on tests/longevity/longevity_test.go:
```diff
@@ -0,0 +1,342 @@
+		}
+
+		// Get next interval of when trigger should happen
+		// This intercal can dynamically change by editing configMap
```
intercal

harsh-px commented on tests/longevity/longevity_test.go:
```diff
@@ -0,0 +1,342 @@
+		// This intercal can dynamically change by editing configMap
+		waitTime, isTriggerEnabled := isTriggerEnabled(triggerType)
+
+		if isTriggerEnabled {
```
can be combined and changed to `if isTriggerEnabled && time.Since(lastInvocationTime) > time.Duration(waitTime) {

harsh-px commented on tests/longevity/longevity_test.go:
```diff
@@ -0,0 +1,342 @@
+				// check no other disruptive trigger is happening at same time
+				takeLock(triggerLoc, triggerType)
+
+				triggerFunc(contexts, triggerEventsChan)
```
just move the logic under takeLock here. Makes the code easier to read. Else only has to co-relate logic in takeLock and the `isDisruptiveTrigger` check below

harsh-px commented on drivers/volume/volume.go:
```diff
@@ -37,6 +37,11 @@ type Image struct {
 	Version string
 }
 
+// Options to pass to APIs
```
update comment for the struct

vinayakshnd commented:
Longevity run after addressing review comments:
https://jenkins.portworx.dev/job/Torpedo/job/tp-vinayak-longevity/76/console

tp-dev-test run:
https://jenkins.portworx.dev/job/Torpedo/job/tp-dev-test/769/console
